src/storage/src/source/persist_source.rs

```diff
@@ -190,3 +194,218 @@ where
+            // `CapabilitySet`, which allows downgrading the contained capabilities to a frontier
+            // but `util::source` gives us both a vanilla `Capability` and a `CapabilitySet` that
+            // we have to keep downgrading because we cannot simply drop the one that we don't
+            // need.
+
+            let mut current_ts = 0;
+
+            move |cap, output| {
+                let mut context = Context::from_waker(&waker);
+                // Bound execution of operator to prevent a single operator from hogging
+                // the CPU if there are many messages to process
+                let timer = Instant::now();
+
+                let mut i = 0;
+
+                while let Poll::Ready(item) = pinned_stream.as_mut().poll_next(&mut context) {
+                    match item {
+                        Some(Ok(batch)) => {
+                            let session_cap = cap.delayed(&current_ts);
+                            let mut session = output.session(&session_cap);
+                            let progress = batch.generate_progress();
+
+                            session.give((i, batch));
+
+                            // Round robin
+                            i = (i + 1) % peers;
+
+                            if let Some(frontier) = progress {
+                                match frontier.into_option() {
+                                    Some(ts) => {
+                                        cap.downgrade(&ts);
+                                        current_ts = ts;
+                                    }
+                                    None => return SourceStatus::Done,
+                                }
+                            }
+                        }
+                        Some(Err::<_, ExternalError>(e)) => {
+                            panic!("unexpected error from persist {e}")
+                        }
+                        None => return SourceStatus::Done,
+                    }
+
+                    if timer.elapsed() > YIELD_INTERVAL {
+                        activator.activate();
+                        break;
+                    }
+                }
+
+                SourceStatus::Alive
+            }
+        });
+
+    let mut builder = OperatorBuilder::new("partitioned reader".to_string(), scope.clone());
+    let dist = |&(i, _): &(usize, ReaderEnrichedHollowBatch<u64>)| i as u64;
+    let mut input = builder.new_input(&inner, Exchange::new(dist));
+    let (mut output, output_stream) = builder.new_output();
+
+    let mut current_ts = 0;
+
+    builder.build_async(
+        scope.clone(),
+        async_op!(|initial_capabilities, _frontiers| {
+            let persist_client = persist_clients
+                .lock()
+                .await
+                .open(metadata.persist_location.clone())
+                .await
+                .expect("could not open persist client")
+                .clone();
+
+            let read = persist_client
```

Comment:
note to self: we'll need to think carefully through the since capabilities of both of these operators, to make sure everything is downgrading correctly

Reply:
If it helps, I think the second operator is morally a `flat_map`, albeit one backed by S3 rather than a function. In that case, as long as a hollow batch is transmitted with a timestamp corresponding to its `lower` frontier, the mostly-vanilla timely idioms should work (hold the capability you receive the hollow batch with, until it has been completely produced).

I'm now reading the code though, and inclined to propose that, unless there are strong negative feels (in which case, no worries). If the *first* operator receives batch descriptions and progress statements, it can forward the batch descriptions to workers, and hold back its capability based on the progress statements. The *second* operators then don't need to do anything about capabilities (frontiers are held back by the first operator holding the capability).

Lmk if this is confusing or wrong. I think it makes a bunch of sense, but also I'm not sure if there are `persist` constraints that make sending a stream of `HollowBatch` things easier than `Enum(HollowBatch, Progress)`.